src/QsCompiler/DocumentationGenerator/Extensions.cs

```diff
@@ -0,0 +1,419 @@
+                String.Join(",", type.TypeItems.ToSyntax())
+            };";
+
+        internal static string ToSyntax(this ResolvedCharacteristics characteristics) =>
+            characteristics.SupportedFunctors.ValueOr(null) switch
+            {
+                null => "",
+                { Count: 0 } => "",
+                var functors => $@" is {String.Join(" + ", 
+                    functors.Select(functor => functor.Tag switch
+                    {
+                        QsFunctor.Tags.Adjoint => "Adj",
+                        QsFunctor.Tags.Controlled => "Ctl"
+                    })
+                )}"
+            };
+
+        internal static string ToSyntax(this QsCallable callable)
+        {
+            var kind = callable.Kind.Tag switch
+            {
+                QsCallableKind.Tags.Function => "function",
+                QsCallableKind.Tags.Operation => "operation",
+                QsCallableKind.Tags.TypeConstructor => "function"
+            };
+            var modifiers = callable.Modifiers.Access.Tag switch
+            {
+                AccessModifier.Tags.DefaultAccess => "",
+                AccessModifier.Tags.Internal => "internal "
+            };
+            var typeParameters = callable.Signature.TypeParameters switch
+            {
+                { Length: 0 } => "",
+                var typeParams => $@"<{
+                    String.Join(", ", typeParams.Select(
+                        param => param switch
+                        {
+                            QsLocalSymbol.ValidName name => $"'{name.Item.Value}",
+                            _ => "{invalid}"
+                        }
+                    ))
+                }>"
+            };
+            var input = callable.ArgumentTuple.ToSyntax();
+            var output = callable.Signature.ReturnType.ToSyntax();
+            var characteristics = callable.Signature.Information.Characteristics.ToSyntax();
+            return $"{modifiers}{kind} {callable.FullName.Name.Value}{typeParameters}{input} : {output}{characteristics}";
+        }
+
+        internal static string MaybeWithSection(this string document, string name, string? contents) =>
+            contents == null || contents.Trim().Length == 0
+            ? document
+            : $"{document}\n\n## {name}\n\n{contents}";
+
+        internal static string WithYamlHeader(this string document, object header) =>
+            $"---\n{new SerializerBuilder().Build().Serialize(header)}---\n{document}";
+
+        internal static bool IsDeprecated(this QsCallable callable, out string? replacement) =>
+            callable.Attributes.IsDeprecated(out replacement);
+
+        internal static bool IsDeprecated(this QsCustomType type, out string? replacement) =>
+            type.Attributes.IsDeprecated(out replacement);
+
+        internal static bool IsDeprecated(this IEnumerable<QsDeclarationAttribute> attributes, [NotNullWhen(true)]  out string? replacement)
```

Comment:
Since the compiler already needs and implements the same logic, please us SymbolResolution.TryFindRedirect instead.